File: content/14-2d-dynamic-programming/word_wrap.js
/**
 * Problem: Word Wrap Problem
 * GeeksforGeeks: https://practice.geeksforgeeks.org/problems/word-wrap/0
 */

function solveWordWrap(nums, k) {
    /**
     * Time Complexity: O(N^2)
     * Space Complexity: O(N)
     */
    const n = nums.length;
    const dp = new Array(n).fill(Infinity);
    
    for (let i = n - 1; i >= 0; i--) {
        let currLen = -1;
        for (let j = i; j < n; j++) {
            currLen += nums[j] + 1;
            if (currLen > k) {
                break;
            }
            
            let cost = 0;
            if (j !== n - 1) {
                cost = Math.pow(k - currLen, 2);
            }
            
            const nextCost = cost + (j + 1 < n ? dp[j + 1] : 0);
            if (nextCost < dp[i]) {
                dp[i] = nextCost;
            }
        }
    }
    
    return dp[0];
}

if (require.main === module) {
    const assert = require('assert');
    assert.strictEqual(solveWordWrap([3, 2, 2, 5], 6), 10);
    assert.strictEqual(solveWordWrap([4, 3], 6), 4);
    console.log("JS: solveWordWrap passed tests!");
}

File: content/14-2d-dynamic-programming/word_wrap.py
"""
Problem: Word Wrap Problem
GeeksforGeeks: https://practice.geeksforgeeks.org/problems/word-wrap/0
"""

def solve_word_wrap(nums: list[int], k: int) -> int:
    """
    Time Complexity: O(N^2)
    Space Complexity: O(N)
    """
    n = len(nums)
    
    # dp[i] represents the minimum cost to wrap words from index i to n-1
    dp = [0] * n
    ans = [0] * n # To track partition (optional, but good)
    
    # If starting word is i, we try to put as many words as possible on the current line
    for i in range(n - 1, -1, -1):
        curr_len = -1
        dp[i] = float('inf')
        
        for j in range(i, n):
            curr_len += nums[j] + 1
            if curr_len > k:
                break
                
            # If we reached the end of the array, the cost is 0 (last line has 0 cost)
            if j == n - 1:
                cost = 0
            else:
                cost = (k - curr_len) ** 2
                
            next_cost = cost + (dp[j + 1] if j + 1 < n else 0)
            if next_cost < dp[i]:
                dp[i] = next_cost
                ans[i] = j
                
    return dp[0]

if __name__ == "__main__":
    assert solve_word_wrap([3, 2, 2, 5], 6) == 10
    assert solve_word_wrap([4, 3], 6) == 4
    print("Python: solve_word_wrap passed tests!")
